db/view/view.cc

```diff
@@ -1775,15 +1790,40 @@ get_view_natural_endpoint(
         }
     }
 
-    SCYLLA_ASSERT(base_endpoints.size() == view_endpoints.size());
     auto base_it = std::find(base_endpoints.begin(), base_endpoints.end(), me);
     if (base_it == base_endpoints.end()) {
         // This node is not a base replica of this key, so we return empty
         // FIXME: This case shouldn't happen, and if it happens, a view update
         // would be lost.
         ++cf_stats.total_view_updates_on_wrong_node;
-        return {};
+        return std::make_pair(std::nullopt, std::nullopt);
+    }
+
+    // The following code handles the case where after a replication factor change
+    // the base replica completed the corresponding tablet migration at a different
+    // time than the view replica.
+    // RF can be changed only by 1 at a time, so the difference between the number
+    // of base and view replicas should be at most 1.
+    SCYLLA_ASSERT(std::abs((int)(base_endpoints.size() - view_endpoints.size())) <= 1);
+    std::optional<locator::host_id> extra_view_endpoint;
+    if (base_endpoints.size() > view_endpoints.size()) {
+        // Because RF changes only affect the last replica in the list, we can simply
+        // ignore the extra base replica. If we are the extra base replica, we should
+        // pair with nothing
+        if (base_it == base_endpoints.end() - 1) {
+            return std::make_pair(std::nullopt, std::nullopt);
+        }
+        base_endpoints.pop_back();
+        ++cf_stats.total_base_view_replicas_mismatch;
+    } else if (base_endpoints.size() < view_endpoints.size()) {
+        // As in the case above, we remove the last view replica from the list.
+        // This replica won't get an update from its paired base replica (as it didn't
+        // finish its tablet migration yet), so we should send an extra update to it now.
+        extra_view_endpoint = std::make_optional(view_endpoints.back());
+        view_endpoints.pop_back();
```

Comment:
won't this overload the poor extra view replica since all base ones will be sending updates to it? but i understand it's a matter of safety (in consistency) and it's only transient, so it's fine I think.

Reply:
It will a bit but that's what's we're doing for pending view replicas already so it should be fine. We're basically waiting for the new base replica to stop being pending instead